SynoCommunity/spksrc
# redesign installer logs

_Motivation:_  Installation scripts using `>> ${INST_LOG}` do not work with DSM7
_Linked issues:_ #4524, #4494, #4527, #4537, #4545, #4153

### Checklist

Documentation that needs update:
- [ ] [CONTRIBUTING.md](https://github.com/SynoCommunity/spksrc/blob/master/CONTRIBUTING.md)
- [ ] [generic service support](https://github.com/SynoCommunity/spksrc/wiki/Service-Support#installer)


### Remarks
It is time to redesign the package installation logs, as appending output with  `>> ${INST_LOG}` is not supported with DSM7.
In DSM7 write access to the new DSM installation log file `/var/log/packages/{package-name.log}` is not allowed. 
With DSM7 all stderr output during installation is written to the logfile by the system and output to stdout is presented in a modal dialog to notify the user.

The purpose of this PR is to realize installation logging independently on the DSM version.

Finally the new function `install_log` is used internally by the generic installer only.
For all packages with custom installer functions (service_*) the only change is to remove the output redirections `>> ${INST_LOG}` and `>> ${INST_LOG} 2>&1`.
All output in the package specific installer functions will be written to the DSM dependent installer log file.

Packages that use `exit 1` and/or user notifications in the installation process by writing to stdout, must use the new validate-functions to do so. User notification is not supported for DSM < 6. 

The new validate functions are:
- validate_preinst
- validate_preunint
- validate_preupgrade

The benefit of negative validation (exit1) is, that the user can navigate in the wizard to fix the installation parameters and try again. This might be `mysql administration password`, `shared folder that must exist`, etc.

_For the rare case a package needs to download or install code before the validation, a validate function must be used and the install_log function can be used there for regular installation logs._

### Installer log file
The install log with `INST_LOG` was introduced for pre DSM 6 packages, where DSM system did not create log file for package installation scripts.
Finally the `/var/log/packages/{package-name}.log` is not new for DSM7 but is provided since DSM6. So this PR modifies all DSM6 packages to use the `/var/log/packages/{package-name}.log` and do not create `/var/package/{package-name}/target/var/{package-name}_install.log` anymore.
The standard for installer log is now `/var/log/packages/{package-name}.log`.
Only for DSM < 6 the `/var/package/{package-name}/target/var/{package-name}_install.log` is still in use.

### Affected Packages
most of the Comments below are for building arch-x64-7.0

✔️: validation of installer log fixed for DSM7 
✔️✔️: validation of installer log fixed for DSM7 and DSM6

fixed:
- remove `>> ${INST_LOG}`
- firewall rules for DSM7 (@publicarray)
- replace `/usr/local/{package-name}`by `/var/packages/{package-name}/target`

not fixed in this PR:
- remove busybox and deprecated user management at installation (DSM 5 to DSM 6 migration is not supported anymore!)
- not all packages use the new var folder with DSM7, some use it only for log and pid of service)
- package warnings like `has not been updated to DSM7 yet` as function ... is no longer supported.
- `SERVICE_WIZARD_SHARE` implementation needs update vor DSM7

| Package | WIP | Comment |
| - | - | - |
| ✔️✔️ beets| | noarch |
| ✔️✔️ borgbackup| |
| ✔️✔️ couchpotatoserver| | noarch |
| ✔️✔️ couchpotatoserver-custom| | noarch |
| ✔️✔️ deluge| #4153 | compile error: expected primary-expression before ‘float’|
| ✔️✔️ demoservice| | noarch, created link to @installer.log does not work on DSM6, but works on DSM7 |
| ✔️✔️ dnscrypt-proxy| | not installable, requires root permissions, DSM6 has logs about missing `localized string` for several languages  |
| ✔️✔️ fish| #4494 | fails to compile for DSM7,  error: ‘__builtin_isnan’ is not a member of ‘std’ |
| ✔️✔️ fishnet| | no wizard to enter `fishnet_key` |
| ✔️✔️ flexget| | service log has all entries doubled |
| ✔️✔️ gateone| | service log has all entries doubled, DSM7: certificate cannot be installed, start fails |
| ✔️✔️ headphones| | noarch |
| ✔️✔️ headphones-custom| | noarch |
| ✔️✔️ homeassistant| |
| ✔️✔️ itools| | not installable, requires root permissions, depends on python module |
| ✔️✔️ lazylibrarian| | noarch |
| ✔️✔️ lidarr| |
| ✔️✔️ minio| |DSM7: migrate SERVICE_EXE (start-stop-daemon) to SERVICE_COMMAND|
| ✔️✔️ mono| |
| ✔️✔️ mosquitto| | mosquitto.conf and *.pid in wrong var folder |
| ✔️✔️ mutt| |
| ✔️✔️ mylar| | datadir in wrong var folder, fails to start on DSM6 |
| ✔️✔️ nzbget| | with multistage validation and and print logfile name to stdout, must not use `su` anymore |
| ✔️✔️ nzbget-testing| | with multistage validation and print logfile name to stdout, must not use `su` anymore, package is obsolete, as `stable / testing (beta)` is provided by the wizard of nzbget package |
| ✔️✔️ nzbhydra| | noarch |
| ✔️✔️ plexpy-custom| noarch |
| ✔️ pyload| |
| ✔️✔️ python3| |
| ✔️✔️ python38| |
| ✔️✔️ rabbitmq| |
| ✔️✔️ radarr| [dsm7-packages](https://github.com/SynoCommunity/spksrc/compare/dsm7-packages) |
| ✔️✔️ rdiff-backup| | usr/local links do not work |
| ✔️✔️ rsnapshot| | noarch, configure: error: rsync not found (DSM 6 and 7)|
| ✔️✔️ rutorrent| |libtorrent:  error: ‘__builtin_isnan’ is not a member of ‘std’|
| ✔️✔️ sabnzbd| #4537 | installation fails, /var/log/messages: failed to alloc _share_, failed to get defualt status for _share_, failed to create _share_, failed to acquire resource after install |
| ✔️✔️ salt-minion| | installation stays `installing...` |
| ✔️✔️ sickbeard-custom| | noarch |
| ✔️✔️ sickrage| | noarch, fails to install: missing apispec==4.0.0 |
| ✔️✔️ sonarr (nzbdrone) | |
| ✔️✔️ stockfish| | web app must not access `/var/services/web/{package-name}` |
| ✔️✔️ syncthing| #4527 | DSM7: migrate SERVICE_EXE (start-stop-daemon) to SERVICE_COMMAND|
| ✔️✔️ synocli-monitor| |
| ✔️✔️ transmission|  [dsm7-packages](https://github.com/SynoCommunity/spksrc/compare/dsm7-packages)  | installation fails, /var/log/messages: failed to alloc _share_, failed to get defualt status for _share_, failed to create _share_, failed to acquire resource after install |
| ✔️✔️ tvheadend| #4545,  [dsm7-packages](https://github.com/SynoCommunity/spksrc/compare/dsm7-packages)  | configure ERROR: ffmpeg development support not found, ffmpeg:  error: ‘__builtin_isnan’ is not a member of ‘std’|
| ✔️✔️ umurmur| | certificate generation failed, /var/packages/umurmur/target/openssl.cnf is missing |
